packages/vite/src/node/optimizer/index.ts

```diff
@@ -359,6 +359,29 @@ export async function loadCachedDepOptimizationMetadata(
 
   const depsCacheDir = getDepsCacheDir(config, ssr)
 
+  const writingPath = path.join(depsCacheDir, '_writing')
+  const maxWaitTime = 500
+  const tryAgainMs = 100
+  let waited = 0
+  let files
+  while (fs.existsSync(writingPath)) {
+    files ??= await fsp.readdir(depsCacheDir)
+    await new Promise((r) => setTimeout(r, tryAgainMs))
+    waited += tryAgainMs
+    if (waited >= maxWaitTime) {
+      const newFiles = await fsp.readdir(depsCacheDir)
+      if (files.length === newFiles.length) {
+        // no new files, outdated deps cache
+        config.logger.info('Outdated deps cache, forcing re-optimization...')
+        await fsp.rm(depsCacheDir, { recursive: true, force: true })
+        return
+      } else {
+        // new files were saved, try again
+        return loadCachedDepOptimizationMetadata(config, ssr, force, asCommand)
+      }
+    }
+  }
```

Comment:
It would be nice to extract this as a function, and some additional comments of:

1. what does the writing file path do?
2. why are we waiting for it?
3. why are we checking the files length?

Took me a while to understand it 😅 

It would also be great to make this non-recursive. If we want to try again, we could extend the `maxWaitTime` manually to continue the loop.